torch/_lobpcg.py

```diff
@@ -13,6 +13,336 @@
+                     poly[..., i] = (a_{i_0}, ..., a{i_n} (==1)), and
+                     poly(x) = poly[..., 0] * zero_power + ... + poly[..., n] * x^n
+
+      x (Tensor): the value (possible batched) to evalate the polynomial `poly` at.
+
+      zero_power (Tensor): the represenation of `x^0`. It is application-specific.
+
+      transition (Callable): the function that accepts some intermediate result `int_val`,
+                             the `x` and a specific polynomial coefficient
+                             `poly[..., k]` for some iteration `k`.
+                             It basically performs one iteration of the Horner's rule
+                             defined as `x * int_val + poly[..., k] * zero_power`.
+                             Note that `zero_power` is not a parameter,
+                             because the step `+ poly[..., k] * zero_power` depends on `x`,
+                             whether it is a vector, a matrix, or something else, so this
+                             functionality is delegated to the user.
+    """
+
+    res = zero_power.clone()
+    for k in range(poly.size(-1) - 2, -1, -1):
+        res = transition(res, x, poly[..., k])
+    return res
+
+def _matrix_polynomial_value(poly, x, zero_power=None):
+    """
+    Evaluates `poly(x)` for the (batched) matrix input `x`.
+    Check out `_polynomial_value` function for more details.
+    """
+
+    # matrix-aware Horner's rule iteration
+    def transition(curr_poly_val, x, poly_coeff):
+        res = x.matmul(curr_poly_val)
+        res.diagonal(dim1=-2, dim2=-1).add_(poly_coeff.unsqueeze(-1))
+        return res
+
+    if zero_power is None:
+        zero_power = torch.eye(x.size(-1), x.size(-1), dtype=x.dtype, device=x.device) \
+            .view(*([1] * len(list(x.shape[:-2]))), x.size(-1), x.size(-1))
+
+    return _polynomial_value(poly, x, zero_power, transition)
+
+def _vector_polynomial_value(poly, x, zero_power=None):
+    """
+    Evaluates `poly(x)` for the (batched) vector input `x`.
+    Check out `_polynomial_value` function for more details.
+    """
+
+    # vector-aware Horner's rule iteration
+    def transition(curr_poly_val, x, poly_coeff):
+        res = torch.addcmul(poly_coeff.unsqueeze(-1), x, curr_poly_val)
+        return res
+
+    if zero_power is None:
+        zero_power = x.new_ones(1).expand(x.shape)
+
+    return _polynomial_value(poly, x, zero_power, transition)
+
+def _symeig_backward_partial_eigenspace(D_grad, U_grad, A, D, U, largest):
+    # compute a projection operator onto an orthogonal subspace spanned by the
+    # columns of U defined as (I - UU^T)
+    Ut = U.transpose(-2, -1).contiguous()
+    proj_U_ortho = -U.matmul(Ut)
+    proj_U_ortho.diagonal(dim1=-2, dim2=-1).add_(1)
+
+    # compute U_ortho, a basis for the orthogonal complement to the span(U),
+    # by projecting a random [..., m, m-k] matrix onto the subspace spanned
+    # by the columns of U.
+    #
+    # fix seed for determinism
```

Comment:
I am not sure why we need this?
The goal is to ensure the backward pass is deterministic?
Given that the random Tensor is not used anywhere else, can we use a "simpler" value to replace it?

Reply:
`U_ortho` is not deterministic, because it is a random projection of a random tensor. When removed, gradcheck says that values do match, but the result is not deterministic.

Reply:
I don't think we ever do this in any other formula so I just want to make sure this is the right choice.
For me the options here are:
- Temporarily change the seed to make the backward fully deterministic. even though we could have a worst case scenario where this projection is very bad for a manually crafted input.
- Don't fix the seed to make the backward non-deterministic. We should update the doc and change the nondet_tol in the gradcheck to have everything working.
- Have a flag to control this like we do for cudnn that the user can set if they actually need the determinism (but the doc should warn that we temporarily change the global random state).

I have to admit I am not sure what is the full impact of changing the global seed here is, but since it is a global state, it can have weird side effects.

cc @gchanan  what do you think about this?

Reply:
As Greg pointed to me, we actually have an API just for that:
You can create a local randomness generator with `gen=torch.Generator(device)` and then pass it to the random function as a kwarg `generator=gen`.
That should ensure determinism as well as avoid issues related to changing the global state.